eslint and prettier CI checks for PRs

To improve our code quality and consistency, I propose we implement Continuous Integration (CI) checks using ESLint and Prettier for all pull requests (PRs).

**ESLint** will help us identify and fix problematic patterns in our JavaScript code, catching potential bugs before they reach production.

**Prettier** will enforce a consistent code style across our project, improving readability and maintainability.


Example of running eslint in the repo at 2951d86f5eca3b200c76b70a255f640dd4174b31


```shell
ev@mac website % npx eslint .

...

/Users/evgeniia/repos/website/src/app/PriceChart.tsx
  203:6  warning  React Hook useEffect has missing dependencies: 'selectDataSource' and 'ws'. Either include them or remove the dependency array  react-hooks/exhaustive-deps

/Users/evgeniia/repos/website/src/app/page.tsx
  10:5  error  Identifier 'initial_staticState' is not in camel case  camelcase

✖ 2 problems (1 error, 1 warning)
```


With the CI checks in place, pull requests with code that does not comply with the standards will not pass the checks and we will be able to see the output right away, no need to weight for a reviewer to pick up some things visible to eslint and prettier.

--- a/.github/workflows/lint.yml
+++ b/.github/workflows/lint.yml
@@ -1,0 +1,27 @@
+name: Lint
+
+on:
+  pull_request:
+    branches:
+      - main
+
+jobs:
+  build:
+    runs-on: ubuntu-latest
+
+    steps:
+      - uses: actions/checkout@v2
+
+      - name: Use Node.js
+        uses: actions/setup-node@v2
+        with:
+          node-version: "20"
+
+      - name: Install Dependencies
+        run: npm ci
+
+      - name: Check Formatting
+        run: npx prettier --check .
+
+      - name: Lint
+        run: npx eslint .

--- a/.prettierignore
+++ b/.prettierignore
@@ -1,0 +1,38 @@
+# copied from ./.gitignore
+# it's possible to run `npx prettier -c --ignore-path .gitignore .` 
+# instead of duplicating the .gitignore file,
+# but duplicating way is more IDE friendly
+
+# dependencies
+/node_modules
+/.pnp
+.pnp.js
+
+# testing
+/coverage
+
+# next.js
+/.next/
+/out/
+
+# production
+/build
+
+# misc
+.DS_Store
+*.pem
+
+# debug
+npm-debug.log*
+yarn-debug.log*
+yarn-error.log*
+
+# local env files
+.env*.local
+
+# vercel
+.vercel
+
+# typescript
+*.tsbuildinfo
+next-env.d.ts

--- a/.vscode/extensions.json
+++ b/.vscode/extensions.json
@@ -1,3 +1,7 @@
 {
-  "recommendations": ["esbenp.prettier-vscode", "bradlc.vscode-tailwindcss"]
+  "recommendations": [
+    "esbenp.prettier-vscode",
+    "bradlc.vscode-tailwindcss",
+    "dbaeumer.vscode-eslint"
+  ]
 }

--- a/.vscode/extensions.json
+++ b/.vscode/extensions.json
@@ -2,6 +2,7 @@
   "recommendations": [
     "esbenp.prettier-vscode",
     "bradlc.vscode-tailwindcss",
-    "dbaeumer.vscode-eslint"
+    "dbaeumer.vscode-eslint",
+    "github.vscode-github-actions"
   ]
 }

--- a/README.md
+++ b/README.md
@@ -12,6 +12,7 @@ npm run dev
 ```
 
 If you'd like to send and receive trades, you will need to:
+
 1. install [Radix development wallet](https://docs-babylon.radixdlt.com/main/getting-started-developers/wallet/wallet-and-connector-installation.html#_install_the_radix_wallet_preview)
 2. install [connector extension](https://docs-babylon.radixdlt.com/main/getting-started-developers/wallet/wallet-and-connector-installation.html#_install_the_connector)
 

--- a/src/app/PriceChart.tsx
+++ b/src/app/PriceChart.tsx
@@ -140,26 +140,21 @@ function convertAlphaDEXData(data: any[]): OHLCVData[] {
 export function PriceChart() {
   //Returns a candlestick chart of the current pair
   const adexState = useContext(AdexStateContext);
-  const [dataSource, setDataSource] = useState<DataSource>(DataSource.BINANCE);
-  const [data, setData] = useState<OHLCVData[]>([]);
+  const [dataSource, setDataSource] = useState<DataSource>(DataSource.ADEX);
+  const [binanceData, setBinanceData] = useState<OHLCVData[]>([]);
 
-  const [ws, setWs] = useState<WebSocket | null>(null);
+  useEffect(() => {
+    // get binance ws and historical data
+    const ws = new WebSocket(
+      "wss://stream.binance.com:9443/ws/btcusdt@kline_1m"
+    );
 
-  async function selectDataSource(dataSource: DataSource) {
-    setData([]);
-    if (ws) {
-      ws.close();
-    }
-
-    switch (dataSource) {
-      case DataSource.BINANCE:
-        const historicalData = await fetchHistoricalData("BTCUSDT", "1m", 1000);
-        setData(historicalData);
-
-        const newWs = new WebSocket(
-          "wss://stream.binance.com:9443/ws/btcusdt@kline_1m"
-        );
-        newWs.onmessage = async (event) => {
+    fetchHistoricalData("BTCUSDT", "1m", 1000)
+      .then((historicalData) => {
+        setBinanceData(historicalData);
+      })
+      .then(() => {
+        ws.onmessage = async (event) => {
           const response = JSON.parse(event.data);
           const {
             t: time,
@@ -178,36 +173,13 @@ export function PriceChart() {
             close: Number(close),
             value: Number(value),
           };
-          setData((prevData) => cleanData([...prevData, newData]));
+          setBinanceData((prevData) => cleanData([...prevData, newData]));
         };
-        setDataSource(dataSource);
-        setWs(newWs);
-        break;
-
-      case DataSource.ADEX:
-        const adexData = convertAlphaDEXData(adexState.currentPairCandlesList);
-        setData(adexData);
-        setDataSource(dataSource);
-        break;
-    }
-  }
-
-  useEffect(() => {
-    selectDataSource(DataSource.ADEX);
-
+      });
     return () => {
-      if (ws) {
-        ws.close();
-      }
+      ws.close();
     };
   }, []);
-
-  useEffect(() => {
-    if (dataSource === DataSource.ADEX) {
-      const adexData = convertAlphaDEXData(adexState.currentPairCandlesList);
-      setData(adexData);
-    }
-  }, [adexState.currentPairCandlesList, dataSource]);
 
   return (
     <div>
@@ -215,7 +187,7 @@ export function PriceChart() {
         className="select w-full max-w-xs bg-gray-200"
         id="source-selector"
         onChange={(e) => {
-          selectDataSource(e.target.value as DataSource);
+          setDataSource(e.target.value as DataSource);
         }}
       >
         <option value={DataSource.ADEX}>AlphaDEX</option>
@@ -239,7 +211,9 @@ export function PriceChart() {
             ))}
           </select>
 
-          <PriceChartCanvas data={data} />
+          <PriceChartCanvas
+            data={convertAlphaDEXData(adexState.currentPairCandlesList)}
+          />
         </div>
       )}
       {dataSource === DataSource.BINANCE && (
@@ -256,7 +230,7 @@ export function PriceChart() {
             </a>
             .
           </p>
-          <PriceChartCanvas data={data} />
+          <PriceChartCanvas data={binanceData} />
         </div>
       )}
     </div>

--- a/src/app/page.tsx
+++ b/src/app/page.tsx
@@ -7,14 +7,14 @@ import { PairInfo } from "./PairInfo";
 
 adex.init(); //Connect to alphadex websocket
 
-let initial_staticState = new adex.StaticState(adex.clientState.internalState);
+let initialStaticState = new adex.StaticState(adex.clientState.internalState);
 
-export const AdexStateContext = createContext(initial_staticState);
+export const AdexStateContext = createContext(initialStaticState);
 
 // more components here: https://daisyui.com/components/
 
 export default function Home() {
-  const [adexState, setAdexState] = useState(initial_staticState);
+  const [adexState, setAdexState] = useState(initialStaticState);
   const [hydrated, setHydrated] = useState(false);
 
   useEffect(() => {

